Sources/Prometheus/PrometheusCollectorRegistry.swift

```diff
@@ -60,21 +60,72 @@ public final class PrometheusCollectorRegistry: Sendable {
     /// For histograms, the buckets are immutable for a MetricFamily once initialized with the first
     /// metric. See also https://github.com/prometheus/OpenMetrics/issues/197.
     private struct MetricFamily<Metric: Sendable & AnyObject>: Sendable {
-        var metricsByLabelSets: [LabelsKey: Metric]
+        private enum State {
+            case labeled([LabelsKey: Metric])
+            case unlabeled(Metric)
+            case empty
+        }
+
+        let metricsByLabelSets: [LabelsKey: Metric]
         let buckets: HistogramBuckets?
         let help: String?
         let metricUnlabeled: Metric?
 
+        private let state: State
+
         init(
             metricsByLabelSets: [LabelsKey: Metric] = [:],
             buckets: HistogramBuckets? = nil,
             help: String? = nil,
             metricUnlabeled: Metric? = nil
         ) {
+            // Validate mutual exclusivity on creation.
+            if metricUnlabeled != nil && !metricsByLabelSets.isEmpty {
+                fatalError("Cannot have both labeled and unlabeled metrics in the same family.")
+            }
+
             self.metricsByLabelSets = metricsByLabelSets
             self.buckets = buckets
             self.help = help
             self.metricUnlabeled = metricUnlabeled
+
+            // Set internal state for validation.
+            if let unlabeled = metricUnlabeled {
+                self.state = .unlabeled(unlabeled)
+            } else if !metricsByLabelSets.isEmpty {
+                self.state = .labeled(metricsByLabelSets)
+            } else {
+                self.state = .empty
+            }
+        }
+
+        func add(metric: Metric, for labels: [(String, String)]) -> MetricFamily<Metric> {
```

Comment:
Minor and doesn’t have to be done, but fyi. In swift often the “add” would be mutating and the not mutating one would be “adding”. But this is fine, just internal api so no need to bikeshed too much